drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1516 @@
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+		cdma_desc->command_flags |= CDMA_CF_INT_SET;
+#endif
+	}
+}
+
+/**
+ * Check a command descriptor transfer complete status in the Cadence NAND controller.
+ *
+ * @param desc_ptr The pointer to the command descriptor structure.
+ * @param params The Cadence NAND parameters structure.
+ * @retval 0 on success or negative error value on failure.
+ */
+static int cdns_transfer_complete(struct cdns_cdma_command_descriptor *desc_ptr,
+				  struct cadence_nand_params *params)
+{
+#ifdef CONFIG_CDNS_NAND_INTERRUPT_SUPPORT
+	uint32_t status = 0;
+
+	k_sem_take(&params->interrupt_sem_t, K_FOREVER);
+	sys_write32(NF_TDEF_TRD_NUM, (params->nand_base + CMD_STATUS_PTR_ADDR));
+	status = sys_read32((params->nand_base + CMD_STAT_CMD_STATUS));
+	if ((status & (BIT(F_CSTAT_COMP)))) {
+		if ((status & (BIT(F_CSTAT_FAIL)))) {
+			LOG_ERR("Cadence status operation failed %s", __func__);
+			return -EIO;
+		}
+	} else {
+		LOG_ERR("Cadence status complete failed %s", __func__);
+		return -EIO;
+	}
+#else
+	ARG_UNUSED(params);
+
+	if (!WAIT_FOR(((desc_ptr->status & (BIT(F_CSTAT_COMP))) != 0), IDLE_TIME_OUT,
+		      k_msleep(1))) {
+		LOG_ERR("Timed out waiting for thread status response");
+		return -ETIMEDOUT;
+	}
+	if ((desc_ptr->status & (BIT(F_CSTAT_FAIL))) != 0) {
+		LOG_ERR("Cadence status operation failed %s", __func__);
+		return -EIO;
+	}
+#endif
+	return 0;
+}
+
+/**
+ * Send a command descriptor to the Cadence NAND controller for execution.
+ *
+ * @param base_address The base address of the Cadence NAND controller.
+ * @param desc_ptr The pointer to the command descriptor.
+ * @param thread The thread number for the execution.
+ * @retval 0 on success or -ETIMEDOUT error value on failure.
+ */
+static int cdns_nand_send(uintptr_t base_address, char *desc_ptr, uint8_t thread)
+{
+	uint64_t desc_address;
+	uint32_t status;
+	int ret;
+
+	desc_address = (uint64_t)desc_ptr;
+
+	ret = cdns_wait_for_thread(base_address, thread);
+
+	if (ret != 0) {
+		return ret;
+	}
+	/* desc_ptr address passing */
+	sys_write32(desc_address & U32_MASK_VAL, (base_address + CDNS_CMD_REG2));
+	sys_write32((desc_address >> 32) & U32_MASK_VAL, (base_address + CDNS_CMD_REG3));
+	/* Thread selection */
+	status = CMD_0_THREAD_POS_SET(thread);
+	/* CDMA Mode selection */
+	status |= CMD_0_C_MODE_SET(CT_CDMA_MODE);
+	/* CMD 0 Reg write*/
+	sys_write32(status, (base_address + CDNS_CMD_REG0));
+	return 0;
+}
+
+static struct cdns_cdma_command_descriptor *cdma_allocate_desc_memory(uint32_t page_count,
+								      uint32_t max_page_desc)
+{
+	struct cdns_cdma_command_descriptor *cdma_desc;
+
+	if (page_count > max_page_desc) {
+		page_count = max_page_desc;
+	}
+
+	cdma_desc = k_malloc(sizeof(struct cdns_cdma_command_descriptor) * page_count);
+	if (cdma_desc == NULL) {
+		LOG_ERR("Memory allocation error occurred %s", __func__);
+		return NULL;
+	}
+
+	memset(cdma_desc, 0, sizeof(struct cdns_cdma_command_descriptor) * page_count);
```

Comment:
IMHO zeroing buffer at allocation is a waste of time, unless hardware assisted, because most of the buffer will be filled in with some stuff. Unless these buffers are passed to something that always requires it zeroed, then fine.

Reply:
removed.